### PR TITLE
Use a single Dockerfile and targets for more efficient building

### PR DESCRIPTION
**Pull request by jashapiro**
In the current implementation of the Docker images with multiple targets, we build all of the images independently, which results in a lot of extra work. The goal of this PR is to streamline things so we build more of the image in just one run, and then reuse layers to keep the related images more efficiently stored and reused.

The way this should work is by using multistage builds and multiple targets. First the full image is built, then the layers of that image are reused to build each of the smaller images by using the targets in the dockerfile. The matrix jobs should therefore run much faster and more efficiently. It also means we only have one Dockerfile to maintain, with less repeated code.

### CODE REVIEW

**Comment by jashapiro** [comment]
This should now be ready to go! It is much more efficient, though not faster in linear time, because not everything runs in parallel. But we rebuild many many fewer times. 

I also updated the dependency-finding script to be more accurate, and to record origins more consistently, which saves on some package redownloading that I was seeing. 

Let me know if you have questions about the organization of the current Dockerfile: I tried to be pretty organized, but there are many stages and targets in one file now!

File: docker/Dockerfile [hunk 1]
```diff
@@ -22,16 +67,30 @@ RUN Rscript -e "proc <- basilisk::basiliskStart(env = zellkonverter::zellkonvert
...
+FROM anndata AS scvi
+LABEL org.opencontainers.image.title "scpcatools-scvi"
+
+COPY requirements_scvi.txt requirements.txt
+RUN pip install --no-cache-dir -r requirements.txt
+
+
+##########################
+# Full image with all dependencies ---------------------------------------------
+FROM scvi AS full
```
[the remark this answers is not in the document]
Comment (jashapiro) [reply docker/Dockerfile:85]: It does not build on those layers, but it does include the packages for Seurat and reports as all of the requirements are in the full `renv.lock` file that is used for the final layer build. There are some tricky ways I might be able to have it copy the needed files from those images, but it didn't seem quite worth it.

[the remark this answers is not in the document]
Comment (allyhawkins) [reply docker/Dockerfile:85]: Okay this makes sense. I just wanted to make sure they got included.

**Review by allyhawkins — APPROVED** [review]
LGTM

**Pull request merged**